### DISCUSSION TITLE
Feature Request: `yield` with return values

### DISCUSSION THREAD

**Discussion by ncthbrt** [dotnet/csharplang#2207] [Language Ideas]
Currently C# supports the use of `yield return SomeValue;` and `yield break;` however it does not support `var returnValue = yield return SomeValue;`

JavaScript generators functions support this form, which turns out to be highly useful for building monadic workflows.

If C# supports `yield return` with return values, it would allow one to build what is effectively F# computation expressions.

A concrete example of where this would be useful would be when designing an Akka actor. Instead of having to create mock actors and simulate responses when unit testing, one could define a stateless monadic actor API which is executed by the Akka runtime, and when testing iterate over the actor, making assertions about requests to send messages to other actors and injecting return values for other API calls. 

The game engine Unity is another example of where this feature would be useful. Unity uses yield for coroutines. Having yield return values would make coroutines more powerful.

Difficulties with implementing this feature request would be that it would be quite difficult to figure out how modify the enumerable API to retain typesafety of return values and that the underlying state machine implementation may no longer work as AST wise it would be a nested clojure.

**Comment by HaloFour** [comment]
Relevant:

https://github.com/dotnet/csharplang/issues/1957

**Comment by ncthbrt** [comment]
Apologies. Missed that proposal.

**Comment by YairHalberstadt** [comment]
I'm afraid I really don't understand this proposal, although I would definitely like to. It would be really useful if you added some code samples!
